### DISCUSSION TITLE
Object cmdset altered after "move_to" character from room

### DISCUSSION THREAD

**Discussion by jamalainm** [evennia/evennia#2320] [Evennia Questions]
I noticed this when I was using the "examine" command as admin.

Somehow I'm changing the cmdset. Below describes the behavior
- If an object is moved from a room to a character through the "get" command, I can no longer "examine" it.
- No feedback is offered through the terminal (log posted below)
- If the object is dropped I can "examine" it again
- If the object is given to another character it still cannot be examined
- If I reload or restart the server the object can be examined wherever it is
- It can be moved from one character to another (i.e. given) and still examined
- It's only when the movement is from a room to a character that it cannot be examined.

I've monkeyed with the commands a typeclasses A LOT, so I'm sure it's my fault.  I have only made changes inside the 'mygame' folder. [Here's my repo](https://github.com/jamalainm/Nova_Tempora)

However:
- This occurs with  the default object, room, and character classes as well as my goofy ones.

# Log
```
21-02-21 08:34:18 Unhandled error in Deferred:
21-02-21 08:34:18 Unhandled Error
	Traceback (most recent call last):
	  File ".../Evennia/evennia/evennia/commands/cmdhandler.py", line 644, in _run_command
	    ret = cmd.func()
	  File ".../Evennia/evennia/evennia/commands/default/building.py", line 2741, in func
	    ).addCallback(get_cmdset_callback)
	  File ".../Evennia/evenv/lib/python3.7/site-packages/twisted/internet/defer.py", line 322, in addCallback
	    callbackKeywords=kw)
	  File ".../Evennia/evenv/lib/python3.7/site-packages/twisted/internet/defer.py", line 311, in addCallbacks
	    self._runCallbacks()
	--- <exception caught here> ---
	  File ".../Evennia/evenv/lib/python3.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File ".../Evennia/evennia/evennia/commands/default/building.py", line 2659, in get_cmdset_callback
	    self.msg(self.format_output(obj, cmdset).strip())
	  File ".../Evennia/evennia/evennia/commands/default/building.py", line 2548, in format_output
	    all_cmdsets = [(cmdset.key, cmdset) for cmdset in current_cmdset.merged_from]
	builtins.AttributeError: 'NoneType' object has no attribute 'merged_from'
```

Thanks for your time!

**Comment by Griatch** [comment]
Hm, you should be able to examine objects also in your inventory. I just tested this on a vanilla install (master branch): 

```
create/drop box
examine box
get box
examine box
```

This works fine. You say this worked the same for default objects, could you give an example of such an object from your repo? It looks like it expects to have a `current_cmdset` on it but can't find one.

**Comment by jamalainm** [comment]
Absolutely! (I have an older install from 2019 that does not show this behavior; I just can't trace where I messed things up this time around for the 2021 install)

## Transcript of creation & examining default object
```
You create a new Object: box.
-----------------------------------------------------------------------------
Name/key: box (#191)
Typeclass: Object (typeclasses.objects.Object)
Location: Limbo (#2)
Home: Limbo (#2)
Locks:
 call:true(); control:id(1) or perm(Admin); delete:id(1) or perm(Admin);
 drop:holds(); edit:perm(Admin); examine:perm(Builder); get:all();
 puppet:pperm(Developer); tell:perm(Admin); view:all()
Stored Cmdset(s):
 
Merged Cmdset(s):
 <Current merged cmdset> (Union prio 101, duplicates: T)
 evennia.commands.cmdset.CmdSet [ExitCmdSet] (Union prio 101, duplicates: T)
Commands available to box (result of Merged CmdSets):
 ad līmen
Persistent attribute(s):
 desc = You see nothing special.
-----------------------------------------------------------------------------
You pick up box.
admin <just typed 'examine box' with no response issued>
You are carrying:
box  You see nothing special.
You drop box.
-----------------------------------------------------------------------------
Name/key: box (#191)
Typeclass: Object (typeclasses.objects.Object)
Location: Limbo (#2)
Home: Limbo (#2)
Locks:
 call:true(); control:id(1) or perm(Admin); delete:id(1) or perm(Admin);
 drop:holds(); edit:perm(Admin); examine:perm(Builder); get:all();
 puppet:pperm(Developer); tell:perm(Admin); view:all()
Stored Cmdset(s):
 
Merged Cmdset(s):
 <Current merged cmdset> (Union prio 101, duplicates: T)
 evennia.commands.cmdset.CmdSet [ExitCmdSet] (Union prio 101, duplicates: T)
Commands available to box (result of Merged CmdSets):
 ad līmen
Persistent attribute(s):
 desc = You see nothing special.
-----------------------------------------------------------------------------
```

## Log of this exercise
```
21-02-21 17:22:02 [SS] Puppet Success: (Caller: admin(account 1), Target: admin, IP: 127.0.0.1).
21-02-21 17:23:10 Unhandled error in Deferred:
21-02-21 17:23:10 Unhandled Error
	Traceback (most recent call last):
	  File ".../Evennia/evennia/evennia/commands/cmdhandler.py", line 644, in _run_command
	    ret = cmd.func()
	  File ".../Evennia/evennia/evennia/commands/default/building.py", line 2741, in func
	    ).addCallback(get_cmdset_callback)
	  File ".../Evennia/evenv/lib/python3.7/site-packages/twisted/internet/defer.py", line 322, in addCallback
	    callbackKeywords=kw)
	  File ".../Evennia/evenv/lib/python3.7/site-packages/twisted/internet/defer.py", line 311, in addCallbacks
	    self._runCallbacks()
	--- <exception caught here> ---
	  File ".../Evennia/evenv/lib/python3.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File ".../Evennia/evennia/evennia/commands/default/building.py", line 2659, in get_cmdset_callback
	    self.msg(self.format_output(obj, cmdset).strip())
	  File ".../Evennia/evennia/evennia/commands/default/building.py", line 2548, in format_output
	    all_cmdsets = [(cmdset.key, cmdset) for cmdset in current_cmdset.merged_from]
	builtins.AttributeError: 'NoneType' object has no attribute 'merged_from'
```

**Comment by davewiththenicehat** [comment]
When I took a look:
https://github.com/jamalainm/Nova_Tempora/commit/8cd7e40643cc3536a6afa3fe172e343fb460cf9a
The changes in commands.iussa_latina.py stood out to me. 

That error is a command set that has nothing in it.

Another method would be to revert to previous builds and see if the issue persists.

I know this may be a little 'thanks captain obvious', don't forget to add this use of examine to your unit tests.

**Reply by jamalainm** [reply]
Thanks, @davewiththenicehat!

You can probably tell from commands.iussa_latina.py that at best I only half-know what I'm doing, so I'm grateful for all the feedback!

Reverting sounds like a good tactic for isolating the issue and you're right about the unit tests, but blergh.

**Reply by Griatch** [reply]
Rather than reverting it *would* still be good to know how a change caused the examine command to stop working - the `examine` command should (ideally) be protected against subtle errors like this.

**Reply by jamalainm** [reply]
I Think i'll try just adding elements from the repo I shared to a new one bit by bit, testing as I go, and hopefully that will help me find out what happened!

I'm really grateful for the ideas a perspectives you all are offering!

**Comment by davewiththenicehat** [comment]
You will get use to unit test writing.
Then after a while you will start using them more then testing in the web console command line.

My thought, is the only way I can finish a large project is if as I add I know I'm not accidently burning bridges behind me.

**Reply by jamalainm** [reply]
Hello again! I've taken your advice to heart, and I thought I would try a fresh install and write tests for each addition as I go.

I'm having a devil of a time figuring out how to implement tests for new typeclasses, though.

I started with just copying the `typeclasses.characters.Character` to `typeclasses.persōnae.Persōna`, without adding any commands or special properties to the typeclass (just renaming `Character(DefaultCharacter)` to `Persōna(DefaultCharacter)`.

I'd eventually like for this to be the default character typeclass, so I added `BASE_CHARACTER_TYPECLASS = "typeclasses.persōnae.Persōna"` to `mygame/server/conf/settings.py`.

Since I wasn't totally sure where to begin testing a new typeclass, I copied `evennia/objects.tests.py` into my local typeclasses directory, where I imported the new typeclass and replaced all instances of `DefaultCharacter` with `Persōna`, and pass `LocalEvenniaTest` to the `DefaultObjectTest` class definition.

This always fails during:
```
def test_get_objs_with_key_and_typeclass(self):
        query = ObjectDB.objects.get_objs_with_key_and_typeclass(
            "Char", "evennia.objects.objects.Persōna" # I've also changed this to 'typeclasses.persōnae.Persōna', which works from the shell
        )
        self.assertEqual(list(query), [self.char1])
```
with the feedback `AssertionError: Lists differ: [] != [Char]`

Presumably this is related to how the EvenniaTest class in `evennia.utils.test_resources` creates test characters?

So far, I've come to the conclusion that I'm going about this entirely wrong, but I can't find much information on testing new typeclasses. Do you have any suggestions on where to start?

**Reply by davewiththenicehat** [reply]
I do know 'typeclasses.persōnae.Persōna' will not work unless you make the Character class for the unit test that persona class.
`evennia.objects.objects.Persōna` does not exist. That's the evennia folder, so your Persona class does not exist there.

The way I do this is, in file maybe `server\test_evennia.py`:
<pre>
<code>"from evennia.objects.tests import *
class TestObjectManager(TestObjectManager):
    # object_typeclass = Object
    character_typeclass = Persōna  # use your class as the character class for characters in this unit test
    # exit_typeclass = Exit
    # room_typeclass = Room
    def test_get_objs_with_key_and_typeclass(self):
        query = ObjectDB.objects.get_objs_with_key_and_typeclass(
            "Char", "typeclasses.persōnae.Persōna"
        )
        self.assertEqual(list(query), [self.char1])
        query = ObjectDB.objects.get_objs_with_key_and_typeclass(
            "Char", "evennia.objects.objects.DefaultObject"
        )
        self.assertFalse(query)
        query = ObjectDB.objects.get_objs_with_key_and_typeclass(
            "NotFound", "typeclasses.persōnae.Persōna"
        )
        self.assertFalse(query)
        query = ObjectDB.objects.get_objs_with_key_and_typeclass(
            "Char", "typeclasses.persōnae.Persōna", candidates=[self.char1, self.char2]
        )
        self.assertEqual(list(query), [self.char1])
"</code>
</pre>
  
There is bit of overriding in that though. You may not want to run built in tests with your custom tests. I bet you would be running 50 or so built in tests. Just making certain interaction with basic object types would work on your server.

Instead you could add `    character_typeclass = Persōna  # use your class as the character class for characters in this unit test` just below the class declaration of your unit test. After importing the class at the top of the script.

**Reply by strikaco** [reply]
There's nothing special about writing unit tests for custom typeclasses; just make sure you're running the test runner with your own settings file:

`evennia test path.to.typeclass.tests --settings settings.py`

**Reply by davewiththenicehat** [reply]
I am very interested in what you are doing. Porting to any language is a project worth learning about.
Just want to make certain to point out a bit of what I have found with Pythons and Evennia unit testing.

BASE_CHARACTER_TYPECLASS is commented with `# Typeclass for character objects linked to an account (fallback)`. I think that works similar to the command set fall backs. If the server starts and the Character class is not useable, if defaults to the class specified for the account object.
Just in case I did double check the EvenniaTest class. It used the standard evennia.objects.objects.DefaultTYPE for each object type. I followed DefaultObject back a few classes, it was not inheriting anything from BASE_CHARACTER_TYPECLASS.

In a standard evennia unit test `self.char1` has a typeclass of `evennia.objects.objects` it gets this class from line 79 of evennia.utils.test_resources.py which is: `character_typeclass = DefaultCharacter`.
I cleaned my example above up. So it is readable now.
Just below the class declaration I have `character_typeclass = Persōna`. I am replacing the class attribute TestObjectManager.character_typeclass with the Persōna class. Now all Characters made in this unit test are that class. Not evennia.objects.objects.DefaultCharacter directly. One thing to keep in mind while testing, almost certainly Persōna is a child of evennia.objects.objects.DefaultCharacter. If you use a function like `utils.inherits_from()` it will find evennia.objects.objects.DefaultCharacter. So expect true because it likely inherits from that class.

In my example above I have `class TestObjectManager(TestObjectManager):`. If you want to import a unit test directly you can. Above this line you would import TestObjectManager you can do this with: `from evennia.objects.tests import *`. If you do, when you run the file you have done this in ALL the unit tests in evennia.objects.tests will be run. I you want to override them, to run something other than the vanilla test you can. I did in the example by having the class inherit the class it is overriding. I than changed the character_type class to Persōna. Than later on changed the test_get_objs_with_key_and_typeclass module to check for the Persōna class not DefaultCharacter.

strikaco is pointing out that you want to use your settings file during your unit tests. This is very very helpful during the tests for server and session. To make certain your settings do not someone create an issue.
Also he mentioned `path.to.typeclass.tests`. You can call a test module directly. To avoid calling all of your unit tests. You can even call a test class within a module to save even more time. As you get a lot of unit tests this saves a load of time.

I wanted to point this out because Evennia is going to change frequently. If you are making a language port it is essentially a wrap of Evennia. The more you can inherit the less you need to change as Evennia changes.

**Reply by jamalainm** [reply]
Awesome! Your suggestions worked perfectly and I've got some other new typeclasses incorporated into the initial round of testing now.

**Comment by jamalainm** [comment]
Well, I don't have a solution, but I did do a [fresh install with its own repo](https://github.com/jamalainm/NovioraTempora) and as much testing as I could think of as a newbie to unit testing.

So far I just have the new typeclasses and Latin commands for get and drop, and the weird "examine" problem has not reemerged.

**Comment by llzzies** [comment]
If you happen to find the problem causer, I'd be very curious to know. I just noticed I have this issuing occurring, as well.

**Reply by jamalainm** [reply]
One thing I've noticed so far: if I reload the server, examine will work on a picked-up object until it is again picked up. Are you experiencing that, too?

**Reply by davewiththenicehat** [reply]
There is a bug report open for the issue currently. The issue is not anything you would have done. This exists in the source code.
https://github.com/evennia/evennia/issues/2328

**Reply by llzzies** [reply]
@jamalainm - yep, same exact situation for me, too. I recently did some modifying of at_traverse() on exits, so I thought maybe there was something I FUBAR'd in there.
